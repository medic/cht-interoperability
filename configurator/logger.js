const {format, createLogger, transports} = require('winston');
const isProduction = process.env.NODE_ENV === 'production';

const logger = createLogger({
  level: !isProduction ? 'debug' : 'info',
  format: format.combine(
    format.errors({stack: true}),
    format.splat(),
    format.simple(),
  ),
  transports: [
    new transports.Console({
      // change level if in dev environment versus production
      format: format.combine(
        // https://github.com/winstonjs/winston/issues/1345
        format(info => {
          info.level = info.level.toUpperCase();
          return info;
        })(),
        format.colorize(),
        format.timestamp({
          format: 'YYYY-MM-DD HH:mm:ss',
        }),
        format.printf(
          info =>
          `${info.timestamp} ${info.level}: ${info.message} ${info.stack ? info.stack : ''}`
          )
      ),
    }),
  ],
});

module.exports = logger;
