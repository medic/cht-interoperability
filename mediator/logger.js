const {format, createLogger, transports} = require('winston');
const isProduction = process.env.NODE_ENV === 'production';

const logger = createLogger({
  level: !isProduction ? 'debug' : 'info',
  format: format.combine(
    format.splat(),
    format.simple()
  ),
  transports: [
    new transports.Console({
      format: format.combine(
        format(info => {
          info.level = info.level.toUpperCase();
          return info;
        })(),
        format.colorize(),
        format.timestamp({
          format: 'YYYY-MM-DD HH:mm:ss',
        }),
        format.printf(
          info =>
          `${info.timestamp} ${info.level}: ${info.message} ${info.stack ? info.stack : ''}`
          )
      ),
    }),
  ],
});

module.exports = logger;
