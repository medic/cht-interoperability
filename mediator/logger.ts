const {format, createLogger, transports} = require('winston');
const isProduction = process.env.NODE_ENV === 'production';

export const logger = createLogger({
  level: !isProduction ? 'debug' : 'info',
  format: format.combine(
    format.splat(),
    format.simple()
  ),
  transports: [
    new transports.Console({
      format: format.combine(
        format((info: { level: string; }) => {
          info.level = info.level.toUpperCase();
          return info;
        })(),
        format.colorize(),
        format.timestamp({
          format: 'YYYY-MM-DD HH:mm:ss',
        }),
        format.printf(
          (          info: { timestamp: any; level: any; message: any; stack: any; }) =>
          `${info.timestamp} ${info.level}: ${info.message} ${info.stack ? info.stack : ''}`
          )
      ),
    }),
  ],
});
