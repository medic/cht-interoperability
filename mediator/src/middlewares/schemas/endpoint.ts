import joi from 'joi';

export const VALID_SYSTEM =
  'http://terminology.hl7.org/CodeSystem/endpoint-connection-type';
export const VALID_CODE = 'hl7-fhir-rest';

const CodingSchema = joi.object({
  system: joi.string().valid(VALID_SYSTEM).required(),
  code: joi.string().valid(VALID_CODE).required(),
});

export const EndpointSchema = joi.object({
  connectionType: CodingSchema.required(),
  identifier: joi.array().length(1).required(),
});
