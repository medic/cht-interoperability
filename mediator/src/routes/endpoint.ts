import { Router } from 'express';
import { validateBodyAgainst } from '../middlewares';
import { createFhirResource, validateFhirResource } from '../utils/fhir';
import { requestHandler } from '../utils/request';

const router = Router();

const resourceType = 'Endpoint';

router.post(
  '/',
  validateBodyAgainst(validateFhirResource(resourceType)),
  requestHandler((req) => createFhirResource({ ...req.body, resourceType }))
);

export default router;
