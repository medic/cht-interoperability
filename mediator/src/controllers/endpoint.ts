import axios from "axios";
import { FHIR } from "../../config";
import { logger } from "../../logger";

const { url: fhirUrl, username, password } = FHIR;

export async function createEndpoint(endpoint: fhir5.Endpoint) {
  try {
    return await axios.post(`${fhirUrl}/Endpoint`, endpoint, {
      auth: { username, password },
    });
  } catch (error: any) {
    logger.error(error);
    return { status: error.status, data: { message: error.data } };
  }
}
