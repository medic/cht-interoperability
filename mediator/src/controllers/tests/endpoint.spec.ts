import axios from "axios";
import { logger } from "../../../logger";
import { createEndpoint } from "../endpoint";

jest.mock("../../../logger");
jest.mock("axios");
const mockAxios = axios as jest.Mocked<typeof axios>;

describe("Endpoint controllers", () => {
  const endpoint: fhir5.Endpoint = {
    resourceType: "Endpoint",
    address: "",
    connectionType: [],
    status: "active",
  };

  describe("createEndpoint", () => {
    it("creates a endpoint when given a valid endpoint resource", async () => {
      const data = { status: 200, data: {} };

      mockAxios.post.mockResolvedValueOnce(data);

      const res = await createEndpoint(endpoint);

      expect(res.data).toBe(data.data);
      expect(res.status).toEqual(data.status);
      expect(mockAxios.post).toHaveBeenCalled();
      expect(mockAxios.post).toMatchSnapshot();
      expect(logger.error).not.toHaveBeenCalled();
    });

    it("resolves error when the endpoint creation fails", async () => {
      const data = { status: 400, data: {} };

      mockAxios.post.mockRejectedValueOnce(data);

      const res = await createEndpoint(endpoint);

      expect(res.status).toBe(data.status);
      expect(logger.error).toHaveBeenCalled();
      expect(mockAxios.post).toHaveBeenCalled();
      expect(res.data).toBe(res.data);
    });
  });
});
