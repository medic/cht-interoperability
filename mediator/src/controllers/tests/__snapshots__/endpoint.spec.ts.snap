// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Endpoint controllers createEndpoint creates a endpoint when given a valid endpoint resource 1`] = `
[MockFunction] {
  "calls": [
    [
      "http://openhim-core:5001/fhir/Endpoint",
      {
        "address": "",
        "connectionType": [],
        "resourceType": "Endpoint",
        "status": "active",
      },
      {
        "auth": {
          "password": "interop-password",
          "username": "interop-client",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
